text/0000-imply-option.md

```diff
@@ -0,0 +1,146 @@
+```
+    /// Construct `Some(T)`, conditionally on a boolean.
+    fn Option::<T>::on_pred(bool, T);
+    /// A lazy equivalent of `on_pred`.
+    fn Option::<F: FnOnce() -> T>::lazy_pred(bool, F);
+```
+Similar in intention and motivation to the `Try` trait for `Result`.
+
+# Motivation
+[motivation]: #motivation
+
+This addition will increase the legibility of code segments and assist in defining the thought processes and motivations of programmers through code. The use cases of this addition are problems which are expressable in the following predicate form:
+```
+    P(x) : Predicate on `x`.
+    F(y) : Function on `y`
+    P(x) -> F(y)
+```
+Or the following Rust pseudocode:
+```
+    if P(x) {
+        Some(F(y))
+    } else {
+        None
+    }
+```
+The outcome of this addition will reduce repeated code which introduces bugs during refactoring and present the thought process of the programmer in a clearer fashion through their code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Option` type is useful when your code may or may not yield a return value.
+Such code may looks similar to this:
+```
+    let x = 0;
+    
+    if x == 0 {
+        Some(x)
+    } else {
+        None
+    }
+```
+However only the `if` branch of this code segment is the important part we're concerned about in our code:
+```
+    if x == 0 {
+        Some(x)
+    }
+```
+But the `else` branch is required for returning `None` value if `x == 0` evaluates to false.
+Fortunately Rusts `Option` type has functionality to get rid of the unecessary code:
+```
+    let x = 0;
+    
+    Option::on_pred(x == 0, x)
+```
+This code has the exact same behaviour as our original `if` statement. Our code is however compressed to a single line and our intentions are just as clear.
+Have you spotted the possible issue with this solution introduces however? What about this code:
+```
+    Option::on_pred(false, foo())
+```
+The above line of code will always return `None` and always throw away the result of `foo()` wasting our precious computing power every time our code needs to return `None`.
+Rust has thought ahead of this problem though:
+```
+    Option::lazy_pred(false, foo)
+```
+`Option`s `lazy_pred` function leverages lazy evaluation by taking a function pointer as its second argument. If its first argument evaluates to `true` it will return `Some(foo())` but if its first argument is `false` it returns `None` without having to run `foo`. This solves the problem presented in our earlier example without sacrificing the advantages it gave us.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This addition is in essence an implementation of the "implies" or "->" operation from First-order logic.
+In predicate logic, for those unfamiliar, the "->" operation has the following truth table:
```

Comment:
This may seem a pedantic point, but the proposed operation has actually very little to do with first-order logic. Predicate logic (first-order or otherwise) generally concerns itself with taking an expression of a certain form (a sentence) and a complex object which defines the semantics of such expressions (a model) and obtaining a truth value. (Generalising this a bit, you could also ask which truth values result from which variable assignments in a formula with free variables. But that's it.)

What is proposed here is, in a sense, an opposite operation: it takes a truth value and returns an object of some other type. (It could be likened to an Iverson bracket, but even that is a somewhat strained analogy.) No such thing exists in first-order logic: implication takes two truth values and returns a truth value, and every predicate and relation takes a tuple of objects and returns a truth value. There's no going from a truth value to some other thing in predicate logic.

I'd remove all mentions of 'predicate logic' from this proposal. It isn't actually helpful even for those who are familiar with it.